.github/workflows/build-and-test.yaml

```diff
@@ -43,56 +47,69 @@ jobs:
         ./scripts/get-runners.py
     outputs:
-      bases: ${{ steps.charm-to-runner.outputs.bases }}
+      platforms: ${{ steps.charm-to-runner.outputs.platforms }}
 
   build:
-    name: 'Build ${{ matrix.arch.name }}-${{ matrix.arch.arch }}'
+    name: 'Build ${{ matrix.platform.charm_name }}-${{ matrix.platform.arch }}'
     strategy:
       fail-fast: false
       matrix:
-        arch: ${{ fromJSON(needs.collect-bases.outputs.bases) }}
+        platform: ${{ fromJSON(needs.collect-platforms.outputs.platforms) }}
     needs:
-      - collect-bases
+      - collect-platforms
       - static-and-unit-tests
     env:
-      PIP_CACHE_DIR: /tmp/cache/pip
-    runs-on: ["self-hosted", "linux", "${{ matrix.arch.arch == 'amd64' && 'X64' || 'ARM64' }}", "jammy", "large"]
+      CRAFT_SHARED_CACHE: /tmp/charmcraft
+    runs-on: ["self-hosted", "linux", "${{ matrix.platform.arch == 'amd64' && 'X64' || 'ARM64' }}", "jammy", "large"]
     timeout-minutes: 60
     steps:
       - name: Checkout
         uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4
-      - name: Install charmcraft
-        run: sudo snap install --classic --channel=latest/stable charmcraft
+      - name: Setup LXD
+        uses: canonical/setup-lxd@54a5806e490d92e207b57183cf111ed54bbdeff4 # v0.1.2
+        with:
+          channel: 5.21/stable
+      - name: Install charmcraft and charmcraftcache
+        run: |
+          sudo snap install --classic --channel=latest/stable charmcraft
+          sudo apt install -y pipx
+          pipx install charmcraftcache
       - name: Cache Deps
         id: cache-charm-deps
         uses: actions/cache@d4323d4df104b026a6aa633fdb11d772146be0bf # v4
         with:
-          path: ${{ env.PIP_CACHE_DIR }}/**
-          key: charm-deps-${{ matrix.arch.arch }}-${{ hashFiles('**/requirements.txt') }}
+          path: ${{ env.CRAFT_SHARED_CACHE }}/**
+          key: charm-deps-${{ matrix.platform.arch }}-${{ hashFiles('**/requirements.txt') }}
       - name: Build charm
         id: build_charm
+        env:
+          VERSIONS: ${{ join(matrix.platform.versions, ' ') }}
         run: |
-            # Make the restored cache dir accessible by root as we run
-            # charmcraft using sudo in order to use destructive-mode
-            if [ -d ${PIP_CACHE_DIR} ]; then
-              sudo chown root ${PIP_CACHE_DIR}
+          for version in $VERSIONS ; do
+            platform=ubuntu@${version}:${{ matrix.platform.arch }}
+            cmd=charmcraft
+            if '${{ inputs.cache }}' ; then
+              cmd=charmcraftcache
             fi
-            sudo pip config set global.cache-dir ${PIP_CACHE_DIR}
-            sudo charmcraft pack --destructive-mode --verbose
-            # Make the created cache dir accessible by user to enable it to be
-            # cached later on
-            sudo chown -R $(whoami) ${PIP_CACHE_DIR}
+            sudo --user "$USER" --preserve-env \
+                  "$cmd" pack \
```

Comment:
Why do we need to run with sudo? This should no longer be needed now that we use LXD again

Reply:
This is not about running with sudo, it's a temporary change to workaround the bug found in snapd, which caused the cache dir was not configured correctly.
https://bugs.launchpad.net/snapd/+bug/2090938
See
https://github.com/canonical/data-platform-workflows/blob/main/.github/workflows/build_charm.yaml#L126-L132

Reply:
See my comment in the PR to anbox-cloud-charms